README.md
# ffxiv-scapy-layer

Main File is ffxiv.py the layer that will end in scapys contrib folder.

OpCodes are beeing fetched during runtime from: https://github.com/karashiiro/FFXIVOpcodes

sample capture file is pcap.py currently configured for europe.

other ranges can be found here: https://is.xivup.com/adv

## usage:
on windows make sure npcap is installed.
and python 3.9 or highter

pip install scapy

start scapy in the folder where ffxiv.py is present.

on the scapy command prompt enter the following:

import ffxiv

sniff(filter="tcp and net (195.82.50.0/24 or 204.2.229.0/24 or 124.150.157.0/24)", prn=lambda x:x.show(), store=0, session=TCPSession)

you shoud now be able to see the FF14 network traffic.


CHANGELOG.md
## [1.10.1](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.10.0...v1.10.1) (2021-10-19)


### Bug Fixes

* added session and encryption detection ([2c40523](https://s2.behead.de/git/ffxiv-scapy-layer/commit/2c40523a8876abb39afeaf0b13447412864b44b8))

# [1.10.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.9.0...v1.10.0) (2021-10-12)


### Features

* working stream decompression ([033a208](https://s2.behead.de/git/ffxiv-scapy-layer/commit/033a2084e20eda056632d6027346acb57071a158))

# [1.9.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.8.0...v1.9.0) (2021-09-27)


### Features

* working deflate of segments ([f79d3bb](https://s2.behead.de/git/ffxiv-scapy-layer/commit/f79d3bbf19a0d6ea4c5c62d812d315fba96e330c))

# [1.8.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.7.0...v1.8.0) (2021-09-26)


### Bug Fixes

* added sample data to test without game ([525e160](https://s2.behead.de/git/ffxiv-scapy-layer/commit/525e160e0386311700f039248ed348fa763f4981))


### Features

* deflate test ([3f422a6](https://s2.behead.de/git/ffxiv-scapy-layer/commit/3f422a6e82b0339ed0fa00691457d5156fbf9169))

# [1.7.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.6.2...v1.7.0) (2021-09-24)


### Features

* added rudimentary group and whisper parser ([66df951](https://s2.behead.de/git/ffxiv-scapy-layer/commit/66df9518335a4bf6ddf3d70af6c947ee3006a3b3))

## [1.6.2](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.6.1...v1.6.2) (2021-09-24)


### Bug Fixes

* ffxiv packet detection ([7966b5f](https://s2.behead.de/git/ffxiv-scapy-layer/commit/7966b5fa25be87de08648d761cae8edb6f8e1168))

## [1.6.1](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.6.0...v1.6.1) (2021-09-21)


### Bug Fixes

* removed all evals ([7a5c65d](https://s2.behead.de/git/ffxiv-scapy-layer/commit/7a5c65d2879a27f9ea4c368a0c0ff887e3439a82))

# [1.6.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.5.0...v1.6.0) (2021-09-21)


### Features

* Implemented ChatHandler ([682c1d9](https://s2.behead.de/git/ffxiv-scapy-layer/commit/682c1d9a247cfb1d7fef9b23f7a0a7b7bd541588))

# [1.5.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.6...v1.5.0) (2021-09-21)


### Features

* added first pytest cases ([6760a4b](https://s2.behead.de/git/ffxiv-scapy-layer/commit/6760a4bb82748447ab36c9a0ac1af8433635c5e1))

## [1.4.6](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.5...v1.4.6) (2021-09-20)


### Bug Fixes

* switched opcode fetch to cdn ([8537f92](https://s2.behead.de/git/ffxiv-scapy-layer/commit/8537f9273abf719c6efae98f8dd88b0161dc1c20))

## [1.4.5](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.4...v1.4.5) (2021-09-20)


### Bug Fixes

* progress ([3a55bf2](https://s2.behead.de/git/ffxiv-scapy-layer/commit/3a55bf2b5ee600b2eecb0dce7bdcbc3c88e8f080))

## [1.4.4](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.3...v1.4.4) (2021-09-18)


### Bug Fixes

* nearly working tcp_reassembly ([7fb3fa9](https://s2.behead.de/git/ffxiv-scapy-layer/commit/7fb3fa9f959aa026da72f3eb12d349656d059fd4))
* pipeline semantic-release requires node 14 ([efb5735](https://s2.behead.de/git/ffxiv-scapy-layer/commit/efb5735a94e7c8e046bb4c62313946bd109c77db))

## [1.4.3](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.2...v1.4.3) (2021-09-16)


### Bug Fixes

* ci ([8115626](https://s2.behead.de/git/ffxiv-scapy-layer/commit/81156260efd53bbcc9ac7902aca63febeda6986d))
* ci again ([8b60049](https://s2.behead.de/git/ffxiv-scapy-layer/commit/8b60049a98f4e8fcd0db0c4f8bb164a035393fef))

## [1.4.2](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.1...v1.4.2) (2021-09-16)


### Bug Fixes

* ci ([6ba2f71](https://s2.behead.de/git/ffxiv-scapy-layer/commit/6ba2f71cbdb0339c57a9e9422fc1ece97b6b03f5))

## [1.4.1](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.4.0...v1.4.1) (2021-09-16)


### Bug Fixes

* ci ([5a2b329](https://s2.behead.de/git/ffxiv-scapy-layer/commit/5a2b329e706465fa783c00bda07f170f720bce76))

# [1.4.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.3.0...v1.4.0) (2021-09-16)


### Features

* a lot of new classes ([d1a6c0e](https://s2.behead.de/git/ffxiv-scapy-layer/commit/d1a6c0e166435e6be51e3c0733af4041a7ccdcc5))

# [1.3.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.2.0...v1.3.0) (2021-09-16)


### Features

* new classes ([11892fd](https://s2.behead.de/git/ffxiv-scapy-layer/commit/11892fd85e33eba914751d34e7d8cde27d64e255))

# [1.2.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.1.0...v1.2.0) (2021-09-16)


### Features

* impl dynamic class loading on known opcodes ([b5db3d9](https://s2.behead.de/git/ffxiv-scapy-layer/commit/b5db3d995158da7644ec811d007e9b44fa4bc6ef))

# [1.1.0](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.0.6...v1.1.0) (2021-09-16)


### Features

* added version batch ([deae2fe](https://s2.behead.de/git/ffxiv-scapy-layer/commit/deae2fe9abc7c5de295f507543de79b1d81f636a))

## [1.0.6](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.0.5...v1.0.6) (2021-09-15)


### Bug Fixes

* ci ([c233353](https://s2.behead.de/git/ffxiv-scapy-layer/commit/c233353fbb98ef9425e17692b79020cdff6e70c0))
* ci ([c74cc63](https://s2.behead.de/git/ffxiv-scapy-layer/commit/c74cc636134ae081d80f19c982df144b1a874269))
* pylint badge ([10e7e3e](https://s2.behead.de/git/ffxiv-scapy-layer/commit/10e7e3e5f6c8cb6c63f96b597492075e0fa17733))

## [1.0.5](https://s2.behead.de/git/ffxiv-scapy-layer/compare/v1.0.4...v1.0.5) (2021-09-15)


### Bug Fixes

* C0301 ([4a51e77](https://s2.behead.de/git/ffxiv-scapy-layer/commit/4a51e775af248763f5b4eb0eef5eaeca40562944))
* ci ([fd8ea76](https://s2.behead.de/git/ffxiv-scapy-layer/commit/fd8ea76140ad463fc49f62946b293f17dd0c3543))
* pylint ([2411063](https://s2.behead.de/git/ffxiv-scapy-layer/commit/2411063b25ca0813c3fe47dd6385c80e386e4a55))
* pylint ([3985426](https://s2.behead.de/git/ffxiv-scapy-layer/commit/3985426044b0b86aa12dc47bca10650aa5f67760))

parser_local.py
def parse(logs):
    #logs = logs.replace(":"," ")
        #LogLine,
        #ChangeZone,
        #ChangePrimaryPlayer,
        #AddCombatant,
        #RemoveCombatant,
        #AddBuff,
        #RemoveBuff,
        #FlyingText,
        #OutgoingAbility,
        #IncomingAbility = 10,
        #PartyList,
        #PlayerStats,
        #CombatantHP,
        #ParsedPartyMember,
        #NetworkStartsCasting = 20,
        #NetworkAbility,
        #NetworkAOEAbility,
        #NetworkCancelAbility,
        #NetworkDoT,
        #NetworkDeath,
        #NetworkBuff,
        #NetworkTargetIcon,
        #NetworkTargetMarker = 29,
        #NetworkBuffRemove,
        #NetworkGauge,
        #NetworkWorld,
        #Network6D,
        #NetworkNameToggle,
        #NetworkTether,
        #NetworkLimitBreak,
        #NetworkEffectResult,
        #NetworkStatusList,
        #NetworkUpdateHp,
        #Settings = 249,
        #Process,
        #Debug,
        #PacketDump,
        #Version,
        #Error,
        #Timer
    
    timestamp = logs[:14]
    command = logs[15:].split(":")
    if command[0] == "00":
        if command[1] == "000a":
            return f"Spieler: {command[2]} sagt: {command[3]}"
        if command[1] == "000b":
            return f"Spieler: {command[2]} schreit: {command[3]}"
        if command[1] == "001d":
            return f"Spieler: {command[2]} emote: {command[3]}"
        if command[1] == "003d":
            return f"NPC: {command[2]} sagt: {command[3]}"
        if command[1] == "0039":
            return f"Status: {command[2]}"
        if command[1] == "2040":
            return f"Award: {command[2]}"
        if command[1] == "0048":
            return f"Inhaltssuche: {command[2]}"
        if command[1] == "082b":
            return f"Cast_success: {command[2]}"
        if command[1] == "08ab":
            return f"Cast: {command[2]}"
        if command[1] == "08ae":
            return f"Buff aquired: {command[2]}"
        if command[1] == "08b0":
            return f"Buff lost: {command[2]}"
        if command[1] == "08ad":
            return f"Regen: {command[2]}"
        if command[1] == "0839":
            return f"Repair: {command[2]}"
        if command[1] == "0843":
            return f"Loot: {command[2]}"
        if command[1] == "08c3":
            return f"Fishing: {command[2]}"
        if command[1] == "0840":
            return f"XP: {command[2]}"
        if command[1] == "0bb9":
            return f"Quest angenommen: {command[2]}"
        
        
        return command

    if command[0] == "01":
        return f"Zone Change: {command[1]}"
    
    if command[0] == "02":
        return f"Player change: {command[1]}"
    
    if command[0] == "03":
        return f"New Player: {command[2]}"
    
    if command[0] == "04":
        return command
    
    if command[0] == "11":
        return f"PartyList: {command}"
        
    if command[0] == "26":
        return command
    
    if command[0] == "1A":
        return f"Real_Buff: {command}"
    
    if command[0] == "1E":
        return command
    
    if command[0] == "14":
        return command[2]
    
    if command[0] == "15":
        return f"Real_Cast: {command[2]} - {command[4]} -> {command[6]}"
    
    if command[0] == "16":
        return command
    
    if command[0] == "17":
        return command
    
    if command[0] == "18":
        return command
    
    if command[0] == "22":
        return command
    
    if command[0] == "25":
        return command
    
    if command[0] == "27":
        return
    
    if command[0] == "0C":
        return f"Player Stats: {command}"
    
    if command[0] == "FB":
        return f"ACT Info: {command[2]}"

    return command